docs(v4): add migration guide from `3.x.x` to `4.x.x` (#3668)
refinedev/refine: documentation/docs/migration-guide.md

import TabItem from '@theme/TabItem';
import CodeBlock from '@theme/CodeBlock';

## **3.0.XX to 4.0.XX**

### Motivation behind breaking changes

TODO

## 🪄 Migrating your project automatically with refine-codemod ✨ (recommended)

[`@refinedev/codemod`][refine-codemod] package handles the breaking changes for your project automatically, without any manual steps. It migrates your project from `3.x.x` to `4.x.x`.

Just `cd` into root folder of your project (where `package.json` is contained) and run this command:

```sh
npx @refinedev/codemod refine3-to-refine4
```

> And it's done. Now your project uses `refine@3.x.x`.

Reviewer: ```suggestion
And it's done. Now your project uses `refine@4.x.x`.
```

## Migrating your project manually

## Updating the packages

Our npm organization has been changed from `@pankod` to `@refinedev`. So you must update your packages to `@refinedev` organization.

```bash
npm uninstall @pankod/refine-core @pankod/refine-antd @pankod/..

npm i @refinedev/core @refinedev/antd @refinedev/..
```

:::caution
You must make this change for all packages that start with `@pankod`.
:::

> 

Reviewer: It will be difficult to do this manually. The `move-to-new-org` codemod is recommended.

useList({
-    metaData,
+    meta,
})

<RefreshButton
-    metaData
+    meta
/>
```

### Resource `options`'s to `meta`

`options` prop of resource is deprecated. Use `meta` prop instead.

```diff
<Refine
    resources={[
        {
            name: "posts",
-           options: {},
+           meta: {},
        },
    ]}
/>
```

### `<ReadyPage>` component is deprecated

## **`@pankod/refine-antd` changes**

### Import changes

All **Ant Design** components re-exported from `@pankod/refine-antd` have been removed. You should import them from `antd` package directly.

```diff
-import { useTable, SaveButton, Button, Form, Input, Select } from "@pankod/refine-antd";

+import { useTable, SaveButton } from "@refinedev/antd";
+import { Button, Form, Input, Select } from "antd";
```

`Icons` are also removed from `@pankod/refine-antd`. You should import them from `@ant-design/icons` directly.

```diff
-import { Icons } from "@pankod/refine-antd";
-const { EditOutlined } = Icons;

+ import { EditOutlined } from "@ant-design/icons";
```

So, you should install `antd` and `@ant-design/icons` packages to access all components and icons.

```bash
npm install antd @ant-design/icons
```

> 

Reviewer: For those who prefer to do it manually, only codemod that makes these changes can be recommended. (antd, mui, mantine, chakra)

Just `cd` into root folder of your project (where `package.json` is contained) and run this command:

```sh
npx @refinedev/codemod refine3-to-refine4
```

And it's done. Now your project uses `refine@4.x.x`.

## Migrating your project manually

## Updating the packages

Our npm organization has been moved from `@pankod` to `@refinedev`. So you must install the packages with the new organization name.

```bash
npm uninstall @pankod/refine-core @pankod/refine-antd @pankod/..

npm i @refinedev/core @refinedev/antd @refinedev/..
```

:::caution
You must make this change for all packages that start with `@pankod`.
:::

## **`@pankod/refine-core` changes**

### `routerProvider`

### `authProvider`

<h3>Motivation behind the changes</h3>

Our motivation behind the changes in `authProvider` prop in **refine**'s v4 is to make it flexible and customizable enough to cover much more cases than before without going down the rabbit hole.

We wanted to create a common interface for the methods of `authProvider` to make it easier to debug and understand the authentication process more easily.

> Previously, the `authProvider` methods returned a resolved or rejected promise rather than a resolved promise with consistent return types. This behavior is not ideal since a rejected promise is typically associated with an error or exceptional case, which is not the case for an authentication failure. This caused confusion for developers and made it harder to debug unexpected behaviors.

Reviewer 2: ```suggestion
Previously, the `authProvider` methods were expected to be resolved on success and rejected on any failure. This behavior wasn't ideal since rejected promise is typically associated with an error or an exceptional case. Considering some expected errors like wrong email/password combination isn't a failure and needs to be propagated to the user, rejected promises caused confusion for developers and made it harder to debug real unexpected behaviors.
```

> `authProvider` methods now returns a resolved promise in all cases, with an object that contains a `success` key. The `success` key indicates whether the operation was successful or not, and an optional `error` key that contains an `Error` object in case of failure.

Reviewer 2: ```suggestion
`authProvider` methods now returns a resolved promise in all cases, with an object containing `success` key. The `success` key indicates whether the operation was successful or not, and an optional `error` key with `Error` object in case of a failure.
```

> Additionally, the auth hooks no longer have default redirection paths. This behavior could be confusing for developers who were not aware of it. By adding a `redirectTo` property to the `authProvider` method's return object, developers can have more control over where the user is redirected after a successful operation.

Reviewer 2: ```suggestion
Additionally, the auth hooks no longer have default redirection paths. This behavior could be confusing for developers who were not aware of it. By adding a `redirectTo` property to the `authProvider` methods' return object explicitly, developers can have more control over where the user is redirected after a successful operation.
```

<h3>Naming changes</h3>

-   `AuthPovider` interface changed to `AuthBindings`.

    ```diff
    - import { AuthProvider } from "@refinedev/core";
    + import { AuthBindings } from "@refinedev/core";

    - const authProvider: AuthProvider = {/* ... */}
    + const authProvider: AuthBindings = {/* ... */}
    ```

> -   `getUserIdentity` method of the `authProvider` changed to `getIdentity`.

Reviewer 2: ```suggestion
-   `authProvider`'s `getUserIdentity` method renamed to `getIdentity`.
```

-   `checkError` method of the `authProvider` changed to `onError`.
-   `checkAuth` method of the `authProvider` changed to `check`.

    ```diff
    const authProvider = {
    -     getUserIdentity,
    +     getIdentity,
    -     useCheckError,
    +     useOnError,
    -     checkAuth,
    +     check,
    }
    ```

-   `useAuthenticated` hook changed to `useIsAuthenticated`.

    ```diff
    - import { useAuthenticated } from "@refinedev/core";
    + import { useIsAuthenticated } from "@refinedev/core";
    ```

<h3>Methods</h3>

#### `login`

> -   `login` method now requires that a promise be resolved instead of rejected, with a return type of `AuthActionResponse`

Reviewer 2: ```suggestion
-   `login` method now requires promises to be resolved in all cases, with a return type of `AuthActionResponse`
```

> -   `login` method no longer rejects promises, you should resolve the promise with a `success` and an `error` value in case of login failure or success.

Reviewer 2: ```suggestion
-   `login` method expects promises to be resolved in all cases. You should always resolve the promise with a `success` key and an additional `error` key in case of a failure.
```

values={[ {label: 'npm', value: 'npm'}, {label: 'pnpm', value: 'pnpm'}, {label: 'yarn', value: 'yarn'} ]}>

<TabItem value="npm">

```bash
npm install @tanstack/react-query
```

</TabItem>

<TabItem value="pnpm">

```bash
pnpm add @tanstack/react-query
```

</TabItem>

<TabItem value="yarn">

```bash
yarn add @tanstack/react-query
```

</TabItem>

</Tabs>

After that, you can import them from `@tanstack/react-query` package directly instead of `@refinedev/core` package.

```diff
- import { QueryClient } from "@refinedev/core";

+ import { QueryClient } from "@tanstack/react-query";
```

### Update `getList` parameters of `dataProvider`

`getList` parameters of `dataProvider` have been updated.

-   `hasPagination` is deprecated. Use `pagination.mode` instead.
-   `sort` is deprecated. Use `sorters` instead.
-   `metaData` is deprecated. Use `meta` instead.

```diff
 export const dataProvider = {
     getList: ({
         resource,
         pagination: {
+            mode: "off" | "server" | "client",
         },
-        hasPagination,
+        sorters,
-        sort,
         filters,
+        meta,
-        metaData,
     }) => {
         ...
     },
 };
```

:::note

> `getList` parameters still have `hasPagination`, `sort`, and `metaData` props. But they are deprecated. You can use them for now. But we recommend you to use `pagination.mode`, `sorters`, and `meta` props instead.

Reviewer 2: ```suggestion
`getList` parameters will still have `hasPagination`, `sort`, and `metaData` props. But they are deprecated. You can use them with v4. But we recommend you to use `pagination.mode`, `sorters`, and `meta` props instead.
```

```sh
npx @pankod/refine-codemod refine3-to-refine4
```

And it's done. Now your project uses `refine@3.x.x`.

## Migrating your project manually

## Updating the packages

Our npm organization has been changed from `@pankod` to `@refinedev`. So you must update your packages to `@refinedev` organization.

```bash
npm uninstall @pankod/refine-core @pankod/refine-antd @pankod/..

npm i @refinedev/refine-core @refinedev/refine-antd @refinedev/..
```

:::caution
You must make this change for all packages that start with `@pankod`.
:::

:::tip
If the operation doesn't work, try removing your `yarn.lock` or `package-lock.json`. If that doesn't help, remove the `node_modules` folder as well and try again.
:::

## **`@pankod/refine-core` changes**

### `routerProvider`

### `authProvider`

### Update `getList` parameters of `dataProvider`

`getList` parameters of `dataProvider` have been updated.

-   `hasPagination` is deprecated. Use `pagination.mode` instead.
-   `sort` is deprecated. Use `sorters` instead.
-   `metaData` is deprecated. Use `meta` instead.

```diff
 export const dataProvider = {
     getList: ({
         resource,
         pagination: {

> +            mode,

Reviewer 3: ```suggestion
+            mode: "off" | "server" | "client",
```

+    resource,
-    sorter,
+    sorters,
})
```

### `useSelect` hook

-   By default, pagination was disabled. If you want to enable pagination, you should set `pagination.mode` prop to `"server"`.

    ```diff
    useSelect({
    -    hasPagination: false,
    +    pagination: {
    +        mode: "server",
    +    },
    })
    ```

-   `sort` prop is deprecated. Use `sorters` prop instead.

    ```diff
    useSelect({
    -    sort,
    +    sorters,
    })
    ```

### `useCustom` hook

`config.sort` prop is deprecated. Use `config.sorters` prop instead.

```diff
useCustom({
    config: {
-       sort,
+       sorters,
    },
})
```

### `metaData` to `meta`

`metaData` is deprecated in all hooks and components. Use `meta` instead.

```diff
useList({
-    metaData,
+    meta,
})

<RefreshButton
-    metaData
+    meta
/>
```

### Resource `options`'s to `meta`

`options` prop of resource is deprecated. Use `meta` prop instead.

```diff
<Refine
    resources={[
        {
            name: "posts",
-           options: {},
+           meta: {},
        },
    ]}
/>
```

### `<ReadyPage>` component is deprecated

## **`@pankod/refine-antd` changes**

### Import changes

> All **Ant Design** components re-exported from `@pankod/refine-antd` have been removed. You should import them from `antd` package directly.

Reviewer 3: Here, the user can be instructed to install the 'antd' package using npm, pnpm, or yarn.

```diff
-import { useTable, SaveButton, Button, Form, Input, Select } from "@pankod/refine-antd";

+import { useTable, SaveButton } from "@pankod/refine-antd";
+import { Button, Form, Input, Select } from "antd";
```

> `Icons` are also removed from `@pankod/refine-antd`. You should import them from `@ant-design/icons` directly.

Reviewer 3: Here, the user can be instructed to install the '@ant-design/icons' package using npm, pnpm, or yarn.

    })
    ```

-   `hasPagination` prop is deprecated. Use `pagination.mode` instead.

    ```diff
    useTable({
    -   hasPagination,
        pagination: {
    +        mode: "off" | "server" | "client",
        },
    })
    ```

-   `initialSorter` and `permanentSorter` props are deprecated. Use `sorters.initial` and `sorters.permanent` instead.

    ```diff
    useTable({
    -    initialSorter,
    -    permanentSorter,
    +    sorters: {
    +        initial,
    +        permanent,
    +    },
    })
    ```

-   `initialFilter`, `permanentFilter`, and `defaultSetFilterBehavior` props are deprecated. Use `filters.initial`, `filters.permanent`, and `filters.defaultBehavior` instead.

    ```diff
    useTable({
    -    initialFilter,
    -    permanentFilter,
    -    defaultSetFilterBehavior,
    +    filters: {
    +        initial,
    +        permanent,
    +        defaultBehavior,
    +    },
    })
    ```

-   `sorter` and `setSorter` return values are deprecated. Use `sorters` and `setSorters` instead.

    ```diff
    const {
    -   sorter,
    +   sorters,
    -   setSorter,
    +   setSorters,
    } = useTable();
    ```

### `useSimpleList` hook

> -   Now `useSimpleList` hook will not accept all of `<List>` component properties So, you can give their props to `<List>` component directly.

Reviewer 3: Could you give an example of the props we removed? You used to do it this way before, but now it's like this.

    ```

-   `hasPagination` prop is deprecated. Use `pagination.mode` instead.

    ```diff
    useSimpleList({
    -   hasPagination,
        pagination: {
    +        mode: "off" | "server" | "client",
        },
    })
    ```

-   `initialSorter` and `permanentSorter` props are deprecated. Use `sorters.initial` and `sorters.permanent` instead.

    ```diff
    useSimpleList({
    -    initialSorter,
    -    permanentSorter,
    +    sorters: {
    +        initial,
    +        permanent,
    +    },
    })
    ```

-   `initialFilter`, `permanentFilter`, and `defaultSetFilterBehavior` props are deprecated. Use `filters.initial`, `filters.permanent`, and `filters.defaultBehavior` instead.

    ```diff
    useSimpleList({
    -    initialFilter,
    -    permanentFilter,
    -    defaultSetFilterBehavior,
    +    filters: {
    +        initial,
    +        permanent,
    +        defaultBehavior,
    +    },
    })
    ```

-   `sorter` and `setSorter` return values are deprecated. Use `sorters` and `setSorters` instead.

    ```diff
    const {
    -   sorter,
    +   sorters,
    -   setSorter,
    +   setSorters,
    } = useSimpleList();
    ```

### `useSelect` hook

> -   By default, pagination was disabled. If you want to enable pagination, you should set `pagination.mode` prop to `"server"`.

Reviewer 3: We can create a tip box here saying 'codemod can do this automatically'

    ```diff
    useSelect({
    -    hasPagination: false,
    +    pagination: {
    +        mode: "server",
    +    },
    })
    ```

-   `sort` prop is deprecated. Use `sorters` prop instead.

    ```diff
    useSelect({
    -    sort,
    +    sorters,
    })
    ```

### `useCheckboxGroup` and `useRadioGroup` hooks

`sort` prop is deprecated. Use `sorters` prop instead.

```diff
useCheckboxGroup({
-    sort,
+    sorters,
})

useRadioGroup({
-    sort,
+    sorters,
})
```

### `useImport` hook

`resourceName` prop is deprecated. Use `resource` prop instead.

```diff
useImport({
-    resourceName,
+    resource,
})
```

> ### `useMenu` hook is deprecated

Reviewer 3: ```suggestion
### `useMenu` hook is removed
```

> `useMenu` hook is deprecated. It will be exported from `@pankod/refine-core` package.

Reviewer 3: ```suggestion
`useMenu` hook is removed. It will be exported from `@pankod/refine-core` package.
```
